// next/dist/server/app-render/debug-channel-server.node.js
/**
 * Node debug channel implementation.
 * Loaded by debug-channel-server.ts when __NEXT_USE_NODE_STREAMS is enabled.
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
0 && (module.exports = {
    createNodeDebugChannel: null,
    createWebDebugChannel: null
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    createNodeDebugChannel: function() {
        return createNodeDebugChannel;
    },
    createWebDebugChannel: function() {
        return _debugchannelserverweb.createWebDebugChannel;
    }
});
const _nodestream = require("node:stream");
const _debugchannelserverweb = require("./debug-channel-server.web");
function createNodeDebugChannel() {
    const readable = new _nodestream.PassThrough();
    // Use a plain Writable instead of exposing the PassThrough directly.
    // React's renderToPipeableStream detects .read() on the debugChannel and
    // enters bidirectional mode, reading its own output back as commands.
    const writable = new _nodestream.Writable({
        write (chunk, _encoding, callback) {
            readable.push(chunk);
            callback();
        },
        final (callback) {
            readable.push(null);
            callback();
        }
    });
    return {
        serverSide: writable,
        clientSide: {
            readable
        }
    };
}
