// next/dist/server/app-render/instant-validation/instant-validation.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
0 && (module.exports = {
    collectStagedSegmentData: null,
    createCombinedPayloadAtDepth: null,
    createCombinedPayloadStream: null,
    discoverValidationDepths: null
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    collectStagedSegmentData: function() {
        return collectStagedSegmentData;
    },
    createCombinedPayloadAtDepth: function() {
        return createCombinedPayloadAtDepth;
    },
    createCombinedPayloadStream: function() {
        return createCombinedPayloadStream;
    },
    discoverValidationDepths: function() {
        return discoverValidationDepths;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _invarianterror = require("../../../shared/lib/invariant-error");
const _stagedrendering = require("../staged-rendering");
const _manifestssingleton = require("../manifests-singleton");
const _apprenderrenderutils = require("../app-render-render-utils");
const _workasyncstorageexternal = require("../work-async-storage.external");
const _prospectiverenderutils = require("../prospective-render-utils");
const _createerrorhandler = require("../create-error-handler");
const _boundary = require("../../../client/components/instant-validation/boundary");
const _boundaryconstants = require("./boundary-constants");
const _appdirmodule = require("../../lib/app-dir-module");
const _parseloadertree = require("../../../shared/lib/router/utils/parse-loader-tree");
const _nodestream = require("node:stream");
const _streamutils = require("./stream-utils");
const _client = require("react-server-dom-webpack/client");
const _segment = require("../../../shared/lib/segment");
const filterStackFrame = process.env.NODE_ENV !== 'production' ? require('../../lib/source-maps').filterStackFrameDEV : undefined;
const findSourceMapURL = process.env.NODE_ENV !== 'production' ? require('../../lib/source-maps').findSourceMapURLDEV : undefined;
const debug = process.env.NEXT_PRIVATE_DEBUG_VALIDATION === '1' ? console.log : undefined;
function traverseRootSeedDataSegments(initialRSCPayload, processSegment) {
    const { flightRouterState, seedData } = getRootDataFromPayload(initialRSCPayload);
    const [rootSegment] = flightRouterState;
    const rootPath = stringifySegment(rootSegment);
    return traverseCacheNodeSegments(rootPath, flightRouterState, seedData, processSegment);
}
function traverseCacheNodeSegments(path, route, seedData, processSegment) {
    processSegment(path, seedData);
    const [_segment, childRoutes] = route;
    const [_node, parallelRoutesData, _loading, _isPartial] = seedData;
    for(const parallelRouteKey in childRoutes){
        const childSeedData = parallelRoutesData[parallelRouteKey];
        if (!childSeedData) {
            throw Object.defineProperty(new _invarianterror.InvariantError(`Got unexpected empty seed data during instant validation`), "__NEXT_ERROR_CODE", {
                value: "E992",
                enumerable: false,
                configurable: true
            });
        }
        const childRoute = childRoutes[parallelRouteKey];
        // NOTE: if this is a __PAGE__ segment, it might have search params appended.
        // Whoever reads from the cache needs to append them as well.
        const [childSegment] = childRoute;
        const childPath = createChildSegmentPath(path, parallelRouteKey, childSegment);
        traverseCacheNodeSegments(childPath, childRoute, childSeedData, processSegment);
    }
}
function createChildSegmentPath(parentPath, parallelRouteKey, segment) {
    const parallelRoutePrefix = parallelRouteKey === 'children' ? '' : `@${encodeURIComponent(parallelRouteKey)}/`;
    return `${parentPath}/${parallelRoutePrefix}${stringifySegment(segment)}`;
}
function stringifySegment(segment) {
    return typeof segment === 'string' ? encodeURIComponent(segment) : encodeURIComponent(segment[0]) + '|' + segment[1] + '|' + segment[2];
}
async function collectStagedSegmentData(ComponentMod, renderFlightStream, fullPageChunks, fullPageDebugChunks, startTime, hasRuntimePrefetch, clientReferenceManifest, createDebugChannel) {
    const debugChannelAbortController = new AbortController();
    const debugStream = fullPageDebugChunks ? (0, _streamutils.createNodeStreamFromChunks)(fullPageDebugChunks, debugChannelAbortController.signal) : null;
    const { stream, controller } = createStagedStreamFromChunks(fullPageChunks);
    stream.on('end', ()=>{
        // When the stream finishes, we have to close the debug stream too,
        // but delay it to avoid "Connection closed." errors.
        setImmediate(()=>debugChannelAbortController.abort());
    });
    // Technically we're just re-encoding, so nothing new should be emitted,
    // but we add an environment name just in case.
    const environmentName = ()=>{
        const currentStage = controller.currentStage;
        switch(currentStage){
            case _stagedrendering.RenderStage.Static:
                return 'Prerender';
            case _stagedrendering.RenderStage.Runtime:
                return hasRuntimePrefetch ? 'Prefetch' : 'Prefetchable';
            case _stagedrendering.RenderStage.Dynamic:
                return 'Server';
            default:
                currentStage;
                throw Object.defineProperty(new _invarianterror.InvariantError(`Invalid render stage: ${currentStage}`), "__NEXT_ERROR_CODE", {
                    value: "E881",
                    enumerable: false,
                    configurable: true
                });
        }
    };
    // Deserialize the payload.
    // NOTE: the stream will initially be in the static stage, so that's as far as we get here.
    // We still expect the outer structure of the payload to be readable in this state.
    const serverConsumerManifest = {
        moduleLoading: null,
        moduleMap: clientReferenceManifest.rscModuleMapping,
        serverModuleMap: (0, _manifestssingleton.getServerModuleMap)()
    };
    const payload = await (0, _client.createFromNodeStream)(stream, serverConsumerManifest, {
        findSourceMapURL,
        debugChannel: debugStream ?? undefined,
        // Do not pass start/end timings - we do not want to omit any debug info.
        startTime: undefined,
        endTime: undefined
    });
    // Deconstruct the payload into separate streams per segment.
    // We have to preserve the stage information for each of them,
    // so that we can later render each segment in any stage we need.
    const { head } = getRootDataFromPayload(payload);
    const segments = new Map();
    traverseRootSeedDataSegments(payload, (segmentPath, seedData)=>{
        segments.set(segmentPath, createSegmentData(seedData));
    });
    const cache = createSegmentCache();
    const pendingTasks = [];
    /** Track when we advance stages so we can pass them as `endTime` later. */ const stageEndTimes = {
        [_stagedrendering.RenderStage.Static]: -1,
        [_stagedrendering.RenderStage.Runtime]: -1
    };
    const renderIntoCacheItem = async (data, cacheEntry)=>{
        const segmentDebugChannel = cacheEntry.debugChunks ? createDebugChannel() : undefined;
        const itemStream = renderFlightStream(ComponentMod, data, clientReferenceManifest.clientModules, {
            filterStackFrame,
            debugChannel: segmentDebugChannel == null ? void 0 : segmentDebugChannel.serverSide,
            environmentName,
            startTime,
            onError (error) {
                const digest = (0, _createerrorhandler.getDigestForWellKnownError)(error);
                if (digest) {
                    return digest;
                }
                // Forward existing digests
                if (error && typeof error === 'object' && 'digest' in error && typeof error.digest === 'string') {
                    return error.digest;
                }
                // We don't need to log the errors because we would have already done that
                // when generating the original Flight stream for the whole page.
                if (process.env.NEXT_DEBUG_BUILD || process.env.__NEXT_VERBOSE_LOGGING) {
                    const workStore = _workasyncstorageexternal.workAsyncStorage.getStore();
                    (0, _prospectiverenderutils.printDebugThrownValueForProspectiveRender)(error, (workStore == null ? void 0 : workStore.route) ?? 'unknown route', _prospectiverenderutils.Phase.InstantValidation);
                }
            }
        });
        await Promise.all([
            // accumulate Flight chunks
            (async ()=>{
                for await (const chunk of itemStream){
                    writeChunk(cacheEntry.chunks, controller.currentStage, chunk);
                }
            })(),
            // accumulate Debug chunks
            segmentDebugChannel && (async ()=>{
                for await (const chunk of segmentDebugChannel.clientSide.readable){
                    cacheEntry.debugChunks.push(chunk);
                }
            })()
        ]);
    };
    await (0, _apprenderrenderutils.runInSequentialTasks)(()=>{
        {
            const headCacheItem = createSegmentCacheItem(!!fullPageDebugChunks);
            cache.head = headCacheItem;
            pendingTasks.push(renderIntoCacheItem(head, headCacheItem));
        }
        for (const [segmentPath, segmentData] of segments){
            const segmentCacheItem = createSegmentCacheItem(!!fullPageDebugChunks);
            cache.segments.set(segmentPath, segmentCacheItem);
            pendingTasks.push(renderIntoCacheItem(segmentData, segmentCacheItem));
        }
    }, ()=>{
        stageEndTimes[_stagedrendering.RenderStage.Static] = performance.now() + performance.timeOrigin;
        controller.advanceStage(_stagedrendering.RenderStage.Runtime);
    }, ()=>{
        stageEndTimes[_stagedrendering.RenderStage.Runtime] = performance.now() + performance.timeOrigin;
        controller.advanceStage(_stagedrendering.RenderStage.Dynamic);
    });
    await Promise.all(pendingTasks);
    return {
        cache,
        payload,
        stageEndTimes
    };
}
/**
 * Turns accumulated stage chunks into a stream.
 * The stream starts out in Static stage, and can be advanced further
 * using the returned controller object.
 * Conceptually, this is similar to how we unblock more content
 * by advancing stages in a regular staged render.
 * */ function createStagedStreamFromChunks(stageChunks) {
    // The successive stages are supersets of one another,
    // so we can index into the dynamic chunks everywhere
    // and just look at the lengths of the Static/Runtime arrays
    const allChunks = stageChunks[_stagedrendering.RenderStage.Dynamic];
    const numStaticChunks = stageChunks[_stagedrendering.RenderStage.Static].length;
    const numRuntimeChunks = stageChunks[_stagedrendering.RenderStage.Runtime].length;
    const numDynamicChunks = stageChunks[_stagedrendering.RenderStage.Dynamic].length;
    let chunkIx = 0;
    let currentStage = _stagedrendering.RenderStage.Static;
    let closed = false;
    function push(chunk) {
        stream.push(chunk);
    }
    function close() {
        closed = true;
        stream.push(null);
    }
    const stream = new _nodestream.Readable({
        read () {
            // Emit static chunks
            for(; chunkIx < numStaticChunks; chunkIx++){
                push(allChunks[chunkIx]);
            }
            // If there's no more chunks after this stage, finish the stream.
            if (chunkIx >= allChunks.length) {
                close();
                return;
            }
        }
    });
    function advanceStage(stage) {
        if (closed) return true;
        switch(stage){
            case _stagedrendering.RenderStage.Runtime:
                {
                    currentStage = _stagedrendering.RenderStage.Runtime;
                    for(; chunkIx < numRuntimeChunks; chunkIx++){
                        push(allChunks[chunkIx]);
                    }
                    break;
                }
            case _stagedrendering.RenderStage.Dynamic:
                {
                    currentStage = _stagedrendering.RenderStage.Dynamic;
                    for(; chunkIx < numDynamicChunks; chunkIx++){
                        push(allChunks[chunkIx]);
                    }
                    break;
                }
            default:
                {
                    stage;
                }
        }
        // If there's no more chunks after this stage, finish the stream.
        if (chunkIx >= allChunks.length) {
            close();
            return true;
        } else {
            return false;
        }
    }
    return {
        stream,
        controller: {
            get currentStage () {
                return currentStage;
            },
            advanceStage
        }
    };
}
function writeChunk(stageChunks, stage, chunk) {
    switch(stage){
        case _stagedrendering.RenderStage.Static:
            {
                stageChunks[_stagedrendering.RenderStage.Static].push(chunk);
            // fallthrough
            }
        case _stagedrendering.RenderStage.Runtime:
            {
                stageChunks[_stagedrendering.RenderStage.Runtime].push(chunk);
            // fallthrough
            }
        case _stagedrendering.RenderStage.Dynamic:
            {
                stageChunks[_stagedrendering.RenderStage.Dynamic].push(chunk);
                break;
            }
        default:
            {
                stage;
            }
    }
}
async function createCombinedPayloadStream(ComponentMod, renderFlightStream, payload, extraChunksAbortController, renderSignal, clientReferenceManifest, startTime, isDebugChannelEnabled, createDebugChannel) {
    // Collect all the chunks so that we're not dependent on timing of the render.
    let isRenderable = true;
    const renderableChunks = [];
    const allChunks = [];
    const debugChunks = isDebugChannelEnabled ? [] : null;
    const debugChannel = isDebugChannelEnabled ? createDebugChannel() : null;
    let streamFinished;
    await (0, _apprenderrenderutils.runInSequentialTasks)(()=>{
        const stream = renderFlightStream(ComponentMod, payload, clientReferenceManifest.clientModules, {
            filterStackFrame,
            debugChannel: debugChannel == null ? void 0 : debugChannel.serverSide,
            startTime,
            onError (error) {
                const digest = (0, _createerrorhandler.getDigestForWellKnownError)(error);
                if (digest) {
                    return digest;
                }
                // Forward existing digests
                if (error && typeof error === 'object' && 'digest' in error && typeof error.digest === 'string') {
                    return error.digest;
                }
                // We don't need to log the errors because we would have already done that
                // when generating the original Flight stream for the whole page.
                if (process.env.NEXT_DEBUG_BUILD || process.env.__NEXT_VERBOSE_LOGGING) {
                    const workStore = _workasyncstorageexternal.workAsyncStorage.getStore();
                    (0, _prospectiverenderutils.printDebugThrownValueForProspectiveRender)(error, (workStore == null ? void 0 : workStore.route) ?? 'unknown route', _prospectiverenderutils.Phase.InstantValidation);
                }
            }
        });
        streamFinished = Promise.all([
            // Accumulate Flight chunks
            (async ()=>{
                for await (const chunk of stream){
                    allChunks.push(chunk);
                    if (isRenderable) {
                        renderableChunks.push(chunk);
                    }
                }
            })(),
            // Accumulate debug chunks
            debugChannel && (async ()=>{
                for await (const chunk of debugChannel.clientSide.readable){
                    debugChunks.push(chunk);
                }
            })()
        ]);
    }, ()=>{
        isRenderable = false;
        extraChunksAbortController.abort();
    });
    await streamFinished;
    return {
        stream: (0, _streamutils.createNodeStreamWithLateRelease)(renderableChunks, allChunks, renderSignal),
        debugStream: debugChunks ? (0, _streamutils.createNodeStreamFromChunks)(debugChunks, renderSignal) : null
    };
}
function getRootDataFromPayload(initialRSCPayload) {
    // FlightDataPath is an unsound type, hence the additional checks.
    const flightDataPaths = initialRSCPayload.f;
    if (flightDataPaths.length !== 1 && flightDataPaths[0].length !== 3) {
        throw Object.defineProperty(new _invarianterror.InvariantError('InitialRSCPayload does not match the expected shape during instant validation.'), "__NEXT_ERROR_CODE", {
            value: "E994",
            enumerable: false,
            configurable: true
        });
    }
    const flightRouterState = flightDataPaths[0][0];
    const seedData = flightDataPaths[0][1];
    // TODO: handle head
    const head = flightDataPaths[0][2];
    return {
        flightRouterState,
        seedData,
        head
    };
}
async function createValidationHead(cache, releaseSignal, clientReferenceManifest, stageEndTimes, stage) {
    const segmentCacheItem = cache.head;
    if (!segmentCacheItem) {
        throw Object.defineProperty(new _invarianterror.InvariantError(`Missing segment data: <head>`), "__NEXT_ERROR_CODE", {
            value: "E1072",
            enumerable: false,
            configurable: true
        });
    }
    return await deserializeFromChunks(segmentCacheItem.chunks[stage], segmentCacheItem.chunks[_stagedrendering.RenderStage.Dynamic], segmentCacheItem.debugChunks, releaseSignal, clientReferenceManifest, {
        startTime: undefined,
        endTime: stageEndTimes[stage]
    });
}
/**
 * Deserializes a (partial possibly partial) RSC stream, given as a chunk-array.
 * If the stream is partial, we'll wait for `releaseSignal` to fire
 * and then complete the deserialization using `allChunks`.
 *
 * This is used to obtain a partially-complete model (that might contain unresolved holes)
 * and then release any late debug info from chunks that came later before we abort the render.
 * */ function deserializeFromChunks(partialChunks, allChunks, debugChunks, releaseSignal, clientReferenceManifest, timings) {
    const debugChannelAbortController = new AbortController();
    const debugStream = debugChunks ? (0, _streamutils.createNodeStreamFromChunks)(debugChunks, debugChannelAbortController.signal) : null;
    const serverConsumerManifest = {
        moduleLoading: null,
        moduleMap: clientReferenceManifest.rscModuleMapping,
        serverModuleMap: (0, _manifestssingleton.getServerModuleMap)()
    };
    const segmentStream = partialChunks.length < allChunks.length ? (0, _streamutils.createNodeStreamWithLateRelease)(partialChunks, allChunks, releaseSignal) : (0, _streamutils.createNodeStreamFromChunks)(partialChunks);
    segmentStream.on('end', ()=>{
        // When the stream finishes, we have to close the debug stream too,
        // but delay it to avoid "Connection closed." errors.
        setImmediate(()=>debugChannelAbortController.abort());
    });
    return (0, _client.createFromNodeStream)(segmentStream, serverConsumerManifest, {
        findSourceMapURL,
        debugChannel: debugStream ?? undefined,
        startTime: timings == null ? void 0 : timings.startTime,
        endTime: timings == null ? void 0 : timings.endTime
    });
}
function createSegmentData(seedData) {
    const [node, _parallelRoutesData, _unused, isPartial, varyParams] = seedData;
    return {
        node,
        isPartial,
        varyParams
    };
}
function getCacheNodeSeedDataFromSegment(data, slots) {
    return [
        data.node,
        slots,
        /* unused (previously `loading`) */ null,
        data.isPartial,
        data.varyParams
    ];
}
function createSegmentCache() {
    return {
        head: null,
        segments: new Map()
    };
}
function createSegmentCacheItem(withDebugChunks) {
    return {
        chunks: {
            [_stagedrendering.RenderStage.Static]: [],
            [_stagedrendering.RenderStage.Runtime]: [],
            [_stagedrendering.RenderStage.Dynamic]: []
        },
        debugChunks: withDebugChunks ? [] : null
    };
}
/**
 * Whether this segment consumes a URL depth level. Each URL depth
 * represents a potential navigation boundary.
 *
 * The root segment ('') consumes depth 0. Regular segments like
 * 'dashboard' consume the next depth — whether or not they have a
 * layout. Route groups, __PAGE__, __DEFAULT__, and /_not-found don't
 * consume a depth — they share the boundary of their parent.
 */ function segmentConsumesURLDepth(segment) {
    // Dynamic segments (tuples) always consume a URL depth.
    if (typeof segment !== 'string') return true;
    // Route groups, pages, defaults, and not-found don't consume a depth.
    if (segment.startsWith(_segment.PAGE_SEGMENT_KEY) || (0, _segment.isGroupSegment)(segment) || segment === _segment.DEFAULT_SEGMENT_KEY || segment === _segment.NOT_FOUND_SEGMENT_KEY) {
        return false;
    }
    // Everything else consumes a depth, including the root segment ''.
    return true;
}
function discoverValidationDepths(loaderTree) {
    const groupDepthsByUrlDepth = [];
    function recordGroupDepth(urlDepth, groupDepth) {
        while(groupDepthsByUrlDepth.length <= urlDepth){
            groupDepthsByUrlDepth.push(0);
        }
        if (groupDepth > groupDepthsByUrlDepth[urlDepth]) {
            groupDepthsByUrlDepth[urlDepth] = groupDepth;
        }
    }
    // urlDepth tracks the index of the current URL-consuming segment.
    // Groups accumulate at the same index. When the next URL segment
    // is reached, it increments the index and resets the group counter.
    // We start at -1 so the root segment '' increments to 0.
    function walk(tree, urlDepth, groupDepth) {
        const segment = tree[0];
        const { parallelRoutes } = (0, _parseloadertree.parseLoaderTree)(tree);
        const consumesDepth = segmentConsumesURLDepth(segment);
        let nextUrlDepth = urlDepth;
        let nextGroupDepth = groupDepth;
        if (consumesDepth) {
            nextUrlDepth = urlDepth + 1;
            nextGroupDepth = 0;
            recordGroupDepth(nextUrlDepth, 0);
        } else if (typeof segment === 'string' && (0, _segment.isGroupSegment)(segment) && segment !== '(__SLOT__)') {
            // Count real route groups but not the synthetic '(__SLOT__)' segment
            // that Next.js inserts for parallel slots. The synthetic group
            // can't be a real navigation boundary.
            nextGroupDepth++;
            recordGroupDepth(urlDepth, nextGroupDepth);
        }
        for(const key in parallelRoutes){
            walk(parallelRoutes[key], nextUrlDepth, nextGroupDepth);
        }
    }
    walk(loaderTree, -1, 0);
    return groupDepthsByUrlDepth;
}
async function createCombinedPayloadAtDepth(initialRSCPayload, cache, initialLoaderTree, getDynamicParamFromSegment, query, depth, groupDepth, releaseSignal, boundaryState, clientReferenceManifest, stageEndTimes, useRuntimeStageForPartialSegments) {
    let hasStaticSegments = false;
    let hasRuntimeSegments = false;
    // Index 0 is reserved for the root config. Slot markers start at 1.
    const slotStacks = [
        null
    ];
    /**
   * When a segment has multiple parallel routes (a fork), wrap each
   * slot's seed data with a slot marker component. The marker's index
   * in the component stack maps to `slotStacks` for per-slot error
   * attribution. Slot markers start at index 1 (index 0 is root).
   */ function wrapSlotsWithMarkers(slots, results) {
        const keys = Object.keys(slots);
        if (keys.length <= 1) return;
        for (const key of keys){
            const slotSeedData = slots[key];
            if (slotSeedData === null) continue;
            const result = results.get(key);
            const markerIndex = slotStacks.length;
            slotStacks.push((result == null ? void 0 : result.createInstantStack) ?? null);
            const markerName = `${_boundaryconstants.INSTANT_SLOT_MARKER_PREFIX}${markerIndex - 1}${_boundaryconstants.INSTANT_SLOT_MARKER_SUFFIX}`;
            const [node, parallelRoutesData, unused, isPartial, varyParams] = slotSeedData;
            slots[key] = [
                // eslint-disable-next-line @next/internal/no-ambiguous-jsx -- bundled in the server layer
                /*#__PURE__*/ (0, _jsxruntime.jsx)(_boundary.SlotMarker, {
                    name: markerName,
                    children: node
                }, "sm"),
                parallelRoutesData,
                unused,
                isPartial,
                varyParams
            ];
        }
    }
    function getSegment(loaderTree) {
        const dynamicParam = getDynamicParamFromSegment(loaderTree);
        if (dynamicParam) {
            return dynamicParam.treeSegment;
        }
        const segment = loaderTree[0];
        return query ? (0, _segment.addSearchParamsIfPageSegment)(segment, query) : segment;
    }
    async function buildSharedTreeSeedData(loaderTree, parentPath, key, urlDepthConsumed, groupDepthConsumed) {
        const { parallelRoutes } = (0, _parseloadertree.parseLoaderTree)(loaderTree);
        const segment = getSegment(loaderTree);
        const path = parentPath === null ? stringifySegment(segment) : createChildSegmentPath(parentPath, key, segment);
        debug == null ? void 0 : debug(`    ${path || '/'} - Dynamic`);
        const segmentCacheItem = cache.segments.get(path);
        if (!segmentCacheItem) {
            throw Object.defineProperty(new _invarianterror.InvariantError(`Missing segment data: ${path}`), "__NEXT_ERROR_CODE", {
                value: "E995",
                enumerable: false,
                configurable: true
            });
        }
        const segmentData = await deserializeFromChunks(segmentCacheItem.chunks[_stagedrendering.RenderStage.Dynamic], segmentCacheItem.chunks[_stagedrendering.RenderStage.Dynamic], segmentCacheItem.debugChunks, releaseSignal, clientReferenceManifest, null);
        const consumesUrlDepth = segmentConsumesURLDepth(segment);
        const isGroup = typeof segment === 'string' && (0, _segment.isGroupSegment)(segment) && segment !== '(__SLOT__)';
        // Advance counters for this segment before the boundary check,
        // mirroring how discoverValidationDepths counts. URL segments
        // increment urlDepthConsumed, groups increment groupDepthConsumed.
        // The synthetic '(__SLOT__)' segment is excluded — it can't be a
        // real navigation boundary.
        let nextUrlDepth = urlDepthConsumed;
        let currentGroupDepth = groupDepthConsumed;
        if (consumesUrlDepth) {
            nextUrlDepth++;
            currentGroupDepth = 0;
        } else if (isGroup) {
            currentGroupDepth++;
        }
        const pastUrlBoundary = nextUrlDepth > depth;
        const isBoundary = pastUrlBoundary && currentGroupDepth >= groupDepth;
        if (isBoundary) {
            debug == null ? void 0 : debug(`    ['${path}' is the boundary (url=${nextUrlDepth}, group=${currentGroupDepth})]`);
            const finalSegmentData = {
                ...segmentData,
                node: // eslint-disable-next-line @next/internal/no-ambiguous-jsx -- bundled in the server layer
                /*#__PURE__*/ (0, _jsxruntime.jsx)(_boundary.PlaceValidationBoundaryBelowThisLevel, {
                    id: path,
                    children: segmentData.node
                }, "c")
            };
            const slots = {};
            const slotResults = new Map();
            let requiresInstantUI = false;
            let createInstantStack = null;
            let bestConfigDepth = -1;
            for(const parallelRouteKey in parallelRoutes){
                const result = await buildNewTreeSeedData(parallelRoutes[parallelRouteKey], path, parallelRouteKey, false, 0 /* segmentDepth */ );
                slotResults.set(parallelRouteKey, result);
                slots[parallelRouteKey] = result.seedData;
                if (result.requiresInstantUI) {
                    requiresInstantUI = true;
                    if (result.configDepth > bestConfigDepth || result.configDepth === bestConfigDepth && parallelRouteKey === 'children') {
                        bestConfigDepth = result.configDepth;
                        createInstantStack = result.createInstantStack;
                    }
                }
            }
            // Only require this boundary to render if the subtree has an
            // instant config. Unconfigured slot subtrees are allowed to not
            // render (e.g. conditionally excluded by a layout).
            if (requiresInstantUI) {
                boundaryState.requiredIds.add(path);
            }
            wrapSlotsWithMarkers(slots, slotResults);
            return {
                seedData: getCacheNodeSeedDataFromSegment(finalSegmentData, slots),
                requiresInstantUI,
                createInstantStack,
                configDepth: bestConfigDepth
            };
        }
        // Not at the boundary yet — keep walking as shared.
        const slots = {};
        const slotResults = new Map();
        let requiresInstantUI = false;
        let createInstantStack = null;
        let bestConfigDepth = -1;
        for(const parallelRouteKey in parallelRoutes){
            const result = await buildSharedTreeSeedData(parallelRoutes[parallelRouteKey], path, parallelRouteKey, nextUrlDepth, currentGroupDepth);
            slotResults.set(parallelRouteKey, result);
            slots[parallelRouteKey] = result.seedData;
            if (result.requiresInstantUI) {
                requiresInstantUI = true;
                if (result.configDepth > bestConfigDepth || result.configDepth === bestConfigDepth && parallelRouteKey === 'children') {
                    bestConfigDepth = result.configDepth;
                    createInstantStack = result.createInstantStack;
                }
            }
        }
        wrapSlotsWithMarkers(slots, slotResults);
        return {
            seedData: getCacheNodeSeedDataFromSegment(segmentData, slots),
            requiresInstantUI,
            createInstantStack,
            configDepth: bestConfigDepth
        };
    }
    async function buildNewTreeSeedData(lt, parentPath, key, isInsideRuntimePrefetch, segmentDepth) {
        const { parallelRoutes } = (0, _parseloadertree.parseLoaderTree)(lt);
        const { mod: layoutOrPageMod } = await (0, _appdirmodule.getLayoutOrPageModule)(lt);
        const segment = getSegment(lt);
        const path = parentPath === null ? stringifySegment(segment) : createChildSegmentPath(parentPath, key, segment);
        let instantConfig = null;
        let prefetchConfig = null;
        let localCreateInstantStack = null;
        if (layoutOrPageMod !== undefined) {
            instantConfig = layoutOrPageMod.unstable_instant ?? null;
            prefetchConfig = layoutOrPageMod.unstable_prefetch ?? null;
            if (instantConfig === true || typeof instantConfig === 'object' && instantConfig !== null) {
                const rawFactory = layoutOrPageMod.__debugCreateInstantConfigStack;
                localCreateInstantStack = typeof rawFactory === 'function' ? rawFactory : null;
            }
        }
        const segmentHasRuntimePrefetch = prefetchConfig === 'force-runtime';
        let childIsInsideRuntimePrefetch = isInsideRuntimePrefetch;
        let stage;
        if (!isInsideRuntimePrefetch) {
            if (segmentHasRuntimePrefetch) {
                stage = _stagedrendering.RenderStage.Runtime;
                childIsInsideRuntimePrefetch = true;
                hasRuntimeSegments = true;
            } else {
                if (useRuntimeStageForPartialSegments) {
                    stage = _stagedrendering.RenderStage.Runtime;
                    hasRuntimeSegments = true;
                } else {
                    stage = _stagedrendering.RenderStage.Static;
                    hasStaticSegments = true;
                }
            }
        } else {
            stage = _stagedrendering.RenderStage.Runtime;
            hasRuntimeSegments = true;
        }
        debug == null ? void 0 : debug(`    ${path || '/'} - ${_stagedrendering.RenderStage[stage]}`);
        const segmentCacheItem = cache.segments.get(path);
        if (!segmentCacheItem) {
            throw Object.defineProperty(new _invarianterror.InvariantError(`Missing segment data: ${path}`), "__NEXT_ERROR_CODE", {
                value: "E995",
                enumerable: false,
                configurable: true
            });
        }
        const segmentData = await deserializeFromChunks(segmentCacheItem.chunks[stage], segmentCacheItem.chunks[_stagedrendering.RenderStage.Dynamic], segmentCacheItem.debugChunks, releaseSignal, clientReferenceManifest, {
            startTime: undefined,
            endTime: stageEndTimes[stage]
        });
        // Build children first, then determine requiresInstantUI.
        const slots = {};
        const slotResults = new Map();
        let childrenRequireInstantUI = false;
        let childCreateInstantStack = null;
        let bestChildConfigDepth = -1;
        for(const parallelRouteKey in parallelRoutes){
            const childSegmentDepth = segmentConsumesURLDepth(segment) ? segmentDepth + 1 : segmentDepth;
            const result = await buildNewTreeSeedData(parallelRoutes[parallelRouteKey], path, parallelRouteKey, childIsInsideRuntimePrefetch, childSegmentDepth);
            slotResults.set(parallelRouteKey, result);
            slots[parallelRouteKey] = result.seedData;
            if (result.requiresInstantUI) {
                childrenRequireInstantUI = true;
                if (result.configDepth > bestChildConfigDepth || result.configDepth === bestChildConfigDepth && parallelRouteKey === 'children') {
                    bestChildConfigDepth = result.configDepth;
                    childCreateInstantStack = result.createInstantStack;
                }
            }
        }
        wrapSlotsWithMarkers(slots, slotResults);
        // Local config takes precedence over children.
        let requiresInstantUI;
        let createInstantStack;
        let configDepth;
        if (instantConfig === false) {
            requiresInstantUI = false;
            createInstantStack = null;
            configDepth = -1;
        } else if (instantConfig === true || typeof instantConfig === 'object' && instantConfig !== null) {
            requiresInstantUI = true;
            createInstantStack = localCreateInstantStack;
            configDepth = segmentDepth;
        } else {
            requiresInstantUI = childrenRequireInstantUI;
            createInstantStack = childCreateInstantStack;
            configDepth = bestChildConfigDepth;
        }
        return {
            seedData: getCacheNodeSeedDataFromSegment(segmentData, slots),
            requiresInstantUI,
            createInstantStack,
            configDepth
        };
    }
    const { seedData, requiresInstantUI, createInstantStack } = await buildSharedTreeSeedData(initialLoaderTree, null, null, 0 /* urlDepthConsumed */ , 0 /* groupDepthConsumed */ );
    if (!requiresInstantUI) {
        return null;
    }
    // Set the root config at index 0. This is the fallback for errors
    // that occur above any fork (no slot marker in the component stack).
    slotStacks[0] = createInstantStack;
    const { flightRouterState } = getRootDataFromPayload(initialRSCPayload);
    const headStage = hasRuntimeSegments ? _stagedrendering.RenderStage.Runtime : _stagedrendering.RenderStage.Static;
    const head = await createValidationHead(cache, releaseSignal, clientReferenceManifest, stageEndTimes, headStage);
    const payload = {
        ...initialRSCPayload,
        f: [
            [
                flightRouterState,
                seedData,
                head
            ]
        ]
    };
    return {
        payload,
        hasAmbiguousErrors: hasStaticSegments,
        slotStacks
    };
}
