// next/dist/client/components/segment-cache/navigation-testing-lock.js
/**
 * Navigation lock for the Instant Navigation Testing API.
 *
 * Manages the in-memory lock (a promise) that gates dynamic data writes
 * during instant navigation captures, and owns all cookie state
 * transitions (pending → captured-MPA, pending → captured-SPA).
 *
 * External actors (Playwright, devtools) set [0] to start a lock scope
 * and delete the cookie to end one. Next.js writes captured values.
 * The CookieStore handler distinguishes them by value: pending = external,
 * captured = self-write (ignored).
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
0 && (module.exports = {
    getPreLockFetch: null,
    globalFetchOverride: null,
    isNavigationLocked: null,
    startListeningForInstantNavigationCookie: null,
    transitionToCapturedSPA: null,
    updateCapturedSPAToTree: null,
    waitForNavigationLockIfActive: null
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    getPreLockFetch: function() {
        return getPreLockFetch;
    },
    globalFetchOverride: function() {
        return globalFetchOverride;
    },
    isNavigationLocked: function() {
        return isNavigationLocked;
    },
    startListeningForInstantNavigationCookie: function() {
        return startListeningForInstantNavigationCookie;
    },
    transitionToCapturedSPA: function() {
        return transitionToCapturedSPA;
    },
    updateCapturedSPAToTree: function() {
        return updateCapturedSPAToTree;
    },
    waitForNavigationLockIfActive: function() {
        return waitForNavigationLockIfActive;
    }
});
const _approuterheaders = require("../app-router-headers");
const _useactionqueue = require("../use-action-queue");
function parseCookieValue(raw) {
    if (raw === '') {
        return 'empty';
    }
    try {
        const parsed = JSON.parse(raw);
        if (Array.isArray(parsed) && parsed.length >= 3) {
            const rawState = parsed[2];
            return rawState === null ? 'mpa' : 'spa';
        }
    } catch  {}
    return 'pending';
}
function writeCookieValue(value) {
    if (typeof cookieStore === 'undefined') {
        return;
    }
    // Read the existing cookie to preserve its attributes (domain, path),
    // then write back with the new value. This updates the same cookie
    // entry that the external actor created, regardless of how it was
    // scoped.
    //
    // Capture the current lockState and compare it in the callback so we
    // only write if the lock we observed at call time is still held. This
    // guards against two races: (a) the scope ended between get and set
    // (lockState is now null), and (b) the scope ended and a new one was
    // acquired in the same gap (lockState is a different object). In
    // either case we must not write — doing so would leak stale state
    // into the next scope or outlive the current one.
    const lockAtCall = lockState;
    cookieStore.get(_approuterheaders.NEXT_INSTANT_TEST_COOKIE).then((existing)=>{
        if (existing && lockState === lockAtCall && lockAtCall !== null) {
            const options = {
                name: _approuterheaders.NEXT_INSTANT_TEST_COOKIE,
                value: JSON.stringify(value),
                path: existing.path ?? '/'
            };
            if (existing.domain) {
                options.domain = existing.domain;
            }
            cookieStore.set(options);
        }
    });
}
let lockState = null;
function getPreLockFetch() {
    return lockState !== null ? lockState.fetch : null;
}
function acquireLock() {
    if (lockState !== null) {
        return;
    }
    let resolve;
    const promise = new Promise((r)=>{
        resolve = r;
    });
    lockState = {
        promise,
        resolve: resolve,
        fetch: window.fetch
    };
    // Install the fetch blocker. We only intercept `window.fetch` for the
    // duration of the lock so that — outside of a testing scope — user-
    // installed overrides of `window.fetch` are untouched.
    if (process.env.__NEXT_EXPOSE_TESTING_API) {
        window.fetch = globalFetchOverride;
    }
}
function releaseLock() {
    if (lockState === null) {
        return;
    }
    // Restore the pre-lock `window.fetch` before resolving the lock promise
    // so any fetches queued on the promise see the restored fetch.
    if (process.env.__NEXT_EXPOSE_TESTING_API) {
        window.fetch = lockState.fetch;
    }
    const { resolve } = lockState;
    lockState = null;
    resolve();
}
function globalFetchOverride(input, init) {
    if (lockState === null) {
        // Lock is not active. Fall through to the global fetch — we reach this
        // only if a caller captured a reference to this function during a lock
        // scope and invoked it after release.
        return fetch(input, init);
    }
    // Block user-initiated fetches until the lock is released, then dispatch
    // through the fetch captured at acquire time. Reading from `lockState`
    // (rather than `window.fetch`) pins to the capture even if `window.fetch`
    // is reassigned after release.
    const currentLock = lockState;
    return currentLock.promise.then(()=>{
        const preLockFetch = currentLock.fetch;
        return preLockFetch(input, init);
    });
}
function startListeningForInstantNavigationCookie() {
    if (process.env.__NEXT_EXPOSE_TESTING_API) {
        // If the server served a static shell, this is an MPA page load
        // while the lock is held. Transition to captured-MPA and acquire.
        if (self.__next_instant_test) {
            if (typeof cookieStore !== 'undefined') {
                // If the cookie was already cleared during the MPA page
                // transition, reload to get the full dynamic page.
                cookieStore.get(_approuterheaders.NEXT_INSTANT_TEST_COOKIE).then((cookie)=>{
                    if (!cookie) {
                        window.location.reload();
                    }
                });
            }
            writeCookieValue([
                1,
                `c${Math.random()}`,
                null
            ]);
            acquireLock();
        }
        if (typeof cookieStore === 'undefined') {
            return;
        }
        cookieStore.addEventListener('change', (event)=>{
            for (const cookie of event.changed){
                if (cookie.name === _approuterheaders.NEXT_INSTANT_TEST_COOKIE) {
                    const state = parseCookieValue(cookie.value ?? '');
                    if (state === 'pending') {
                        // External actor starting a new lock scope.
                        if (lockState !== null) {
                            releaseLock();
                        }
                        acquireLock();
                    }
                    // Captured value (our own transition) or empty. Ignore.
                    return;
                }
            }
            for (const cookie of event.deleted){
                if (cookie.name === _approuterheaders.NEXT_INSTANT_TEST_COOKIE) {
                    releaseLock();
                    (0, _useactionqueue.refreshOnInstantNavigationUnlock)();
                    return;
                }
            }
        });
    }
}
function transitionToCapturedSPA(fromTree, toTree) {
    if (process.env.__NEXT_EXPOSE_TESTING_API) {
        writeCookieValue([
            1,
            `c${Math.random()}`,
            {
                from: fromTree,
                to: toTree
            }
        ]);
    }
}
function updateCapturedSPAToTree(fromTree, toTree) {
    if (process.env.__NEXT_EXPOSE_TESTING_API) {
        writeCookieValue([
            1,
            `c${Math.random()}`,
            {
                from: fromTree,
                to: toTree
            }
        ]);
    }
}
function isNavigationLocked() {
    if (process.env.__NEXT_EXPOSE_TESTING_API) {
        if (lockState !== null) {
            return true;
        }
        // If `lockState` is null, fall back to reading the test cookie
        // synchronously from `document.cookie`. This accounts for a small race
        // between `cookieStore.set(...)` and its corresponding `change` event.
        // During that gap `lockState` is still null even though the cookie
        // indicates a new lock scope is starting.
        if (typeof document === 'undefined') {
            return false;
        }
        const allCookies = document.cookie;
        if (!allCookies.includes(_approuterheaders.NEXT_INSTANT_TEST_COOKIE)) {
            // Fast bail-out: in almost every navigation the test cookie is not
            // set at all.
            return false;
        }
        const target = _approuterheaders.NEXT_INSTANT_TEST_COOKIE + '=';
        for (const segment of allCookies.split(';')){
            const trimmed = segment.trim();
            if (trimmed.startsWith(target) && parseCookieValue(trimmed.slice(target.length)) === 'pending') {
                // The cookie was set by an external actor but the change event was not
                // yet dispatched. Acquire the lock synchronously.
                acquireLock();
                return true;
            }
        }
    }
    return false;
}
async function waitForNavigationLockIfActive() {
    if (process.env.__NEXT_EXPOSE_TESTING_API) {
        if (lockState !== null) {
            await lockState.promise;
        }
    }
}

if ((typeof exports.default === 'function' || (typeof exports.default === 'object' && exports.default !== null)) && typeof exports.default.__esModule === 'undefined') {
  Object.defineProperty(exports.default, '__esModule', { value: true });
  Object.assign(exports.default, exports);
  module.exports = exports.default;
}
